core/src/main/scala/org/apache/spark/h2o/utils/H2OSchemaUtils.scala

```diff
@@ -70,23 +72,174 @@ object H2OSchemaUtils {
+    val flatSchemaIndexes = flatSchema.fields.map(_.name).zipWithIndex.toMap
+    val originalSchema = df.schema
+    df.map[Row]{ r: Row =>
+      val result = ArrayBuffer.fill[Any](numberOfColumns)(null)
+      val fillMethod = fillBuffer(flatSchemaIndexes, result, r)(_, _, _)
+      originalSchema.fields.zipWithIndex.foreach{ case (f, i) => fillMethod(f, i, None)}
+      Row.fromSeq(result)
+    }
+  }
+
+  private def fillBuffer
+      (flatSchemaIndexes: Map[String, Int], buffer: ArrayBuffer[Any], row: Row)
+      (field: StructField, index: Int, prefix: Option[String] = None): Unit = {
+    val StructField(name, dataType, _, _) = field
+    val qualifiedName = getQualifiedName(prefix, name)
+    if(!row.isNullAt(index)) {
+      dataType match {
+        case BinaryType =>
+          fillArray(flatSchemaIndexes, buffer, row, ByteType, index, qualifiedName)
+        case MapType(_, valueType, containsNull) =>
+          val map = row.getMap[Any, Any](index)
+          val subRow = Row.fromSeq(map.values.toSeq)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          map.keys.zipWithIndex.foreach{ case (k, i) =>
+            val mapField = StructField(k.toString, valueType, containsNull)
+            fillMethod(mapField, i, Some(qualifiedName))
+          }
+        case ArrayType(elementType, _) =>
+          fillArray(flatSchemaIndexes, buffer, row, elementType, index, qualifiedName)
+        case StructType(fields) =>
+          val subRow = row.getStruct(index)
+          val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+          fields.zipWithIndex.foreach { case (f, i) => fillMethod(f, i, Some(qualifiedName)) }
+        case dt => buffer(flatSchemaIndexes(qualifiedName)) = row.get(index)
+      }
+    }
+  }
+
+  private def fillArray(
+      flatSchemaIndexes: Map[String, Int],
+      buffer: ArrayBuffer[Any],
+      row: Row,
+      elementType: DataType,
+      index: Int,
+      qualifiedName: String){
+    val seq = row.getSeq[Any](index)
+    val subRow = Row.fromSeq(seq)
+    val fillMethod = fillBuffer(flatSchemaIndexes, buffer, subRow)(_, _, _)
+    (0 until seq.size).foreach{ i =>
+      val arrayField = StructField(i.toString, elementType)
+      fillMethod(arrayField, i, Some(qualifiedName))
+    }
+  }
+
+  def flattenSchema(df: DataFrame): StructType = {
+    val originalSchema = df.schema
+    val fields = df.rdd.map[Seq[FieldWithOrder]] { row: Row =>
+        originalSchema.fields.zipWithIndex.foldLeft(Seq.empty[FieldWithOrder]) {
+          case (acc, (field, index)) => acc ++ flattenField(field, row, index, index :: Nil)
+        }
+      }
+      .reduce { (first, second) =>
+        def convertToMap(collection: Seq[FieldWithOrder]) = collection.map(f => f.order -> f.field).toMap
+        val firstMap = convertToMap(first)
+        val secondMap = convertToMap(second)
+        val pathItemOrdering = new Ordering[Any] {
```

Comment:
Can we convert the names of field to string on the nested conversion methods so we don't have to specify this ordering? Also Seq[Any] could become Seq[String] as they represent just names

Reply:
I'm quite not happy of `Seq[Any]` as well. This collection contains indexes of fields for each level of nesting to ensure that order of columns won't change after flattening.
E.g.: `{index for level 0} :: {index for level 1} :: {index for level 2} :: Nil`. I wanted to to use `Seq[Int]` but for `MapType` I have to use keys instead of indexes. 

It seems that conversion of number would break ordering. E.g.: `"10"` would be before `"1"`.

Any thoughts how to get rid of `Seq[Any]` and preserve ordering of columns?